### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by vchinnakotla**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+.idea_modules/
+
+# JIRA plugin
+atlassian-ide-plugin.xml
+
+# Crashlytics plugin (for Android Studio and IntelliJ)
+com_crashlytics_export_strings.xml
+crashlytics.properties
+crashlytics-build.properties
+fabric.properties
```
Comment (joseph-wortmann) [inline .gitignore]: Add `.vscode` to gitignore

File: README.md [hunk 2]
```diff
@@ -1,2 +1,30 @@
...
+"ec2:DescribeLaunchTemplates",
+"autoscaling:DescribeAutoScalingGroups",
+"autoscaling:DescribeLaunchConfigurations"
+
+## Request Syntax
+{}
+
+## Response Syntax
+[
+    [
```
Comment (joseph-wortmann) [inline README.md]: Change the reponse to a list of maps, where each map has ami-id and snapshot-id as keys.

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# CHANGES BELOW HERE ARE MADE AT YOUR OWN RISK
+##############################################
+
+setup(
+    name=lambda_name,
+
+    version=lambda_version,
+
+    description=lambda_description,
+    long_description=long_description,
```
Comment (joseph-wortmann) [inline setup.py]: long_description needs to be defined, or you need to change this to point to lambda_description

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        'License :: '+lambda_license,
+
+        # Specify the Python versions you support here. In particular, ensure
+        # that you indicate whether you support Python 2, Python 3 or both.
+        'Programming Language :: Python :: 3.7',
+    ],
+
+    keywords=lambda_keywords,
+
+    install_requires=filter(lambda x : x not in ['boto3', 'botocore'], lambda_install_requires),
```
Comment (joseph-wortmann) [inline setup.py]: change this to `[x for x in lambda_install_requires if x not in ['boto3', 'botocore']]`. What you have will fail silently in Python3.7; `filter` returns a list in 2.7, but a `Filter` object in 3.7.

In the future, use the cookiecutter template that I built and you won't have this issue

File: README.md [hunk 5]
```diff
@@ -1,2 +1,30 @@
 # ami-cleaner
```
Comment (joseph-wortmann) [inline README.md]: Change the README.md to a README.rst please. RST is a better markup language, and what I am using across the board now (that Github supports RST files)

File: src/lambda_function/function.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+ec2_client = boto3.client('ec2')
+autoscaling_client = boto3.client('autoscaling')
+
+OLDER_THAN_DAYS = os.environ['OLDER_THAN_DAYS']
+
+launch_configurations_dump = autoscaling_client.describe_launch_configurations()
```
Comment (joseph-wortmann) [inline src/lambda_function/function.py]: Do these inside the method(s). The results they provide can change rapidly in some accounts, and doing them at the top level will cache the results across lambda function runs.

File: src/lambda_function/function.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+def handler(event, context):
+    response = ami_list(ami_info_dump)
+	#Response contains AMI ID's and there corresponding Snapshots which are removed
+    return response
+
+def ami_list(ami_list):
+    response = []
+    for items in ami_list['Images']:
+        is_expired = expire(items['ImageId'], items['CreationDate'])
+        if is_expired != None :
```
Comment (joseph-wortmann) [inline src/lambda_function/function.py]: This should be `if is_expired:`. The check for `None` is unnecessary

File: src/lambda_function/function.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            response.append(is_expired)
+    return response	
+	
+	    
+def expire(ami_id, creation_date):
+    ami_ids = []
+    date_now = datetime.datetime.now().date()
+    creation_date_only = datetime.datetime.strptime(creation_date, "%Y-%m-%dT%H:%M:%S.%fZ").date()
+
+    if (date_now - creation_date_only).days > int(OLDER_THAN_DAYS):
```
Comment (joseph-wortmann) [inline src/lambda_function/function.py]: Cast `OLDER_THAN_DAYS` value to an int when you define it above in the globals

File: src/lambda_function/function.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    for items in ami_list['Images']:
+        is_expired = expire(items['ImageId'], items['CreationDate'])
+        if is_expired != None : 
+            response.append(is_expired)
+    return response	
+	
+	    
+def expire(ami_id, creation_date):
+    ami_ids = []
+    date_now = datetime.datetime.now().date()
```
Comment (joseph-wortmann) [inline src/lambda_function/function.py]: Make sure that you are using `utcnow`. All AMI dates are UTC

File: src/lambda_function/function.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    creation_date_only = datetime.datetime.strptime(creation_date, "%Y-%m-%dT%H:%M:%S.%fZ").date()
+
+    if (date_now - creation_date_only).days > int(OLDER_THAN_DAYS):
+        if(check_launch_configurations(ami_id) and check_launch_templates(ami_id)):
+            logger.info('Will Deregister AMI {}'.format(ami_id))
+            ami_ids.append(ami_id)
+            deregister_data = deregister_image(ami_id)
+            return ami_ids, deregister_data
```
Comment (joseph-wortmann) [inline src/lambda_function/function.py]: Suggest that you return a `dict` here rather than a `tuple`

**Review by joseph-wortmann — COMMENTED** [review]
Very well done. See inlined changes requested

**Comment by vchinnakotla** [comment]
@joseph-wortmann requested changes are made, please review. If this is released, I will layer up the terraform

**Comment by vchinnakotla** [comment]
Please review, requested changes are made.

**Comment by joseph-wortmann** [comment]
I never reviewed this @vchinnakotla. Is it still appropriate?

**Comment by vchinnakotla** [comment]
No, this repo can be archived.

**Pull request closed without merging**